packages-exp/app-check-exp/src/internal-api.ts

```diff
@@ -0,0 +1,293 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FirebaseApp } from '@firebase/app-exp';
+import {
+  AppCheckTokenResult,
+  AppCheckTokenListener,
+  AppCheckTokenInternal
+} from './types';
+import { getDebugState, getState, setState } from './state';
+import { TOKEN_REFRESH_TIME } from './constants';
+import { Refresher } from './proactive-refresh';
+import { ensureActivated } from './util';
+import { exchangeToken, getExchangeDebugTokenRequest } from './client';
+import { writeTokenToStorage, readTokenFromStorage } from './storage';
+import { getDebugToken, isDebugMode } from './debug';
+import { base64 } from '@firebase/util';
+import { logger } from './logger';
+import { Provider } from '@firebase/component';
+
+// Initial hardcoded value agreed upon across platforms for initial launch.
+// Format left open for possible dynamic error values and other fields in the future.
+export const defaultTokenErrorData = { error: 'UNKNOWN_ERROR' };
+
+/**
+ * Stringify and base64 encode token error data.
+ *
+ * @param tokenError Error data, currently hardcoded.
+ */
+export function formatDummyToken(
+  tokenErrorData: Record<string, string>
+): string {
+  return base64.encodeString(
+    JSON.stringify(tokenErrorData),
+    /* webSafe= */ false
+  );
+}
+
+/**
+ * This function always resolves.
+ * The result will contain an error field if there is any error.
+ * In case there is an error, the token field in the result will be populated with a dummy value
+ */
+export async function getToken(
+  app: FirebaseApp,
+  platformLoggerProvider: Provider<'platform-logger'>,
+  forceRefresh = false
+): Promise<AppCheckTokenResult> {
+  ensureActivated(app);
+  /**
+   * DEBUG MODE
+   * return the debug token directly
+   */
+  if (isDebugMode()) {
+    const tokenFromDebugExchange: AppCheckTokenInternal = await exchangeToken(
+      getExchangeDebugTokenRequest(app, await getDebugToken()),
+      platformLoggerProvider
+    );
+    return { token: tokenFromDebugExchange.token };
+  }
+
+  const state = getState(app);
+
+  let token: AppCheckTokenInternal | undefined = state.token;
+  let error: Error | undefined = undefined;
+
+  /**
+   * try to load token from indexedDB if it's the first time this function is called
+   */
+  if (!token) {
+    // readTokenFromStorage() always resolves. In case of an error, it resolves with `undefined`.
+    const cachedToken = await readTokenFromStorage(app);
+    if (cachedToken && isValid(cachedToken)) {
+      token = cachedToken;
+
+      setState(app, { ...state, token });
+      // notify all listeners with the cached token
+      notifyTokenListeners(app, { token: token.token });
+    }
+  }
+
+  // return the cached token if it's valid
+  if (!forceRefresh && token && isValid(token)) {
+    return {
+      token: token.token
+    };
+  }
+
+  /**
+   * request a new token
+   */
+  try {
+    if (state.provider) {
```

Comment:
I don't think it's possible to not have a provider, so can we use ! to assert it is available, and add a comment to clarify?

Reply:
Done.